src/main/scala/sample/stream/GroupLogFile.scala

```diff
@@ -29,13 +30,18 @@ object GroupLogFile {
         case other                  => "OTHER"
       }.
       // write lines of each group to a separate file
-      foreach {
+      map {
         case (level, groupFlow) =>
           val output = new PrintWriter(new FileOutputStream(s"target/log-$level.txt"), true)
           // close resource when the group stream is completed
           // foreach returns a future that we can key the close() off of
-          groupFlow.foreach(line => output.println(line)).onComplete(_ => Try(output.close()))
+          val future = groupFlow.foreach(line => output.println(line))
+          future.onComplete { _ => Try(output.close()) }
```

Comment:
this should be future.andThen. If you just return the original future below then the completion of that future might happen earlier than the onComplete callback. By using andThen you establish a proper happens-before relation.